docs/CODING_CONVENTIONS.md

```diff
@@ -0,0 +1,259 @@
+        Error::IoError(e)
+    }
+}
+```
+
+#### Error enums should include a `__Nonexhaustive` variant
+
+Any enum-based error type _should_ include a non-exhaustive error variant.
+
+This prevents consumers of our API from relying on pattern-matching which is not future proof against added error
+variants.
+
+###### Example
+
+Implementing a non-exhaustive variant is done like this:
+
+```rust
+pub enum Error {
+    PermissionDenied,
+    NotConnected,
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+```
+
+This informs users of the API to include a catch-all arm when matching against the error:
+
+```rust
+match e {
+    PermissionDenied => { /*  */ },
+    NotConnected => { /*  */ },
+    _ => { /*  */ },
+}
+```
+
+This pattern guards against any matches being non-exhaustive in case new error variants are added in the future.
+
+#### Use `amethyst_error::Error` for compositional APIs
+
+APIs which composes results from multiple crates _should_ use `amethyst_error::Error`.
+This is a generic error type which is capable of boxing any error and annotate it with debugging information.
+
+This must be used when defining APIs which composes errors generically, like with traits.
+This is also required when composing errors from other crates, since It is currently the only mechanism available to
+communicate backtraces.
+
+###### Do
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// no dependencies
+mod c {
+    use amethyst_error::Error;
+
+    pub trait Example {
+        fn foo() -> Result<u32, Error>;
+    }
+}
+```
+
+###### Don't
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::A(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::B(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// depends on `a` and `b`.
+mod c {
+    use crate::{a, b};
+
+    pub enum Error {
+        A(a::error::Error),
```

Comment:
I think the problematic part of this example is that each crate has its own error type and it seems like it would make more sense to define an associated `Error` type.

Reply:
This specific example tries to communicate the cyclical nature of defining "correct" compositional error types in generic APIs. I'd be open to adding more.

I definitely agree that an associated error type is the best solution in a simple scenario. But I've tried to do this in for example `PrefabData` (I encourage you to take a spin!) and you encounter problems like:

* An exploding number of error variants.
* The current inability to communicate backtraces except through `amethyst_error::Error`.
* Difficult scenarios where you need to compose N generic errors that I'm not even sure how to solve.

From this experience, my recommendation would still stand to default towards using a generic error type. But definitely try to attempt and refactor later if possible. I'd be happy to revisit once more things are standardized though.

Reply:
Okay, thanks for the detailed explanation. I won't block this PR any further, please just resolve my last comment.

Reply:
I actually tried having an associated type on `PrefabData` for errors, but as you say it quickly leads to spaghetti code, and loss of valuable information.